### PR TITLE
db: Improve the way database inserts are done

### PR DESCRIPTION
**Pull request by erikd**

### CODE REVIEW

File: cardano-db/src/Cardano/Db/Insert.hs [hunk 1]
```diff
@@ -142,21 +172,70 @@ data DbInsertException
...
+
+    fieldNames, placeholders :: [Text]
+    (fieldNames, placeholders) =
+      unzip (Util.mkInsertPlaceholders (entityDef (Proxy @record)) escape)
+
+    exceptHandler :: SqlError -> ReaderT SqlBackend m a
+    exceptHandler e =
+      liftIO $ throwIO (DbInsertException vtype e)
+
+escape :: DBName -> Text
```
Comment (kderme) [inline cardano-db/src/Cardano/Db/Insert.hs:235]: A pity this is not exported

File: cardano-db/src/Cardano/Db/Insert.hs [hunk 2]
```diff
@@ -142,21 +172,70 @@ data DbInsertException
...
+
+    values :: [PersistValue]
+    values = pvalues ++ [head pvalues]
+
+    pvalues :: [PersistValue]
+    pvalues = map toPersistValue (toPersistFields record)
+
+    fieldNames, placeholders :: [Text]
+    (fieldNames, placeholders) =
+      unzip (Util.mkInsertPlaceholders (entityDef (Proxy @record)) escape)
```
Comment (kderme) [inline cardano-db/src/Cardano/Db/Insert.hs:229]: Does this work? I see persistent uses qualified names (see `escapeF` instead of `escape`.

**Review by kderme — DISMISSED** [review]
LGTM. Only minor improvements

File: cardano-db/src/Cardano/Db/Insert.hs [hunk 3]
```diff
@@ -142,21 +172,70 @@ data DbInsertException
...
+        , MonadIO m
+        , PersistEntity record
+        , PersistRecordBackend record SqlBackend
+        )
+    => String -> record -> ReaderT SqlBackend m (Key record)
+insertCheckUnique vtype record = do
+    res <- handle exceptHandler $ rawSql query values
+    case res of
+      [ident] -> pure ident
+      _other -> error $ mconcat [ "insertByReturnKey2: Inserting ", vtype, " failed with ", show res ]
```
[the remark this answers is not in the document]
Comment (erikd) [reply cardano-db/src/Cardano/Db/Insert.hs:206]: Yeah, print string from an earlier version of this fix.  Unfortunately strings are not compiler checked :trollface:

File: cardano-db/src/Cardano/Db/Insert.hs (hunk 2 above)
Comment (erikd) [reply cardano-db/src/Cardano/Db/Insert.hs:229]: `escapeF` has type `FieldNameDB -> Text` and is not in the current release of `persistent` that we are using.

my version of `escape` has the type `DBName -> Text`, the types fit and the function makes sense.

File: cardano-db/src/Cardano/Db/Insert.hs (hunk 1 above)
Comment (erikd) [reply cardano-db/src/Cardano/Db/Insert.hs:235]: Yes, and it seems it is changing between `persistent-2.11.0.1` and the next release.

File: cardano-db/src/Cardano/Db/Insert.hs [hunk 4]
```diff
@@ -142,21 +172,70 @@ data DbInsertException
...
+  where
+    query :: Text
+    query =
+      Text.concat
+        [ "INSERT INTO "
+        , unDBName (entityDB . entityDef $ Just record)
+        , " (", Util.commaSeparated fieldNames
+        , ") VALUES (", Util.commaSeparated placeholders
+        , ") ON CONFLICT ON CONSTRAINT "
+        , unDBName (uniqueDBName . head . entityUniques . entityDef $ Just record)
```
[the remark this answers is not in the document]
Comment (erikd) [reply cardano-db/src/Cardano/Db/Insert.hs:216]: I think I actually need `onlyOneUniqueDef` which [unfortunately is not exported](https://github.com/yesodweb/persistent/issues/1194). I will cargo cult that in for now.

**Comment by erikd** [comment]
For some reason Hydra is not feeding its result back to GH even though the build was good.

Merging this anyway.

**Pull request merged**